src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,328 @@
+        workerID = (int) tb.getLong(0, i);
+      } else {
+        throw new DbException("WorkerID must be of type INT or LONG");
+      }
+      Preconditions.checkState(workerID > 0, "WorkerID must be > 0");
+      Preconditions.checkState(!workerInfo.containsKey(workerID),
+          "Duplicate WorkerIDs");
+
+      int tupleCount;
+      if (col1Type == Type.INT_TYPE) {
+        tupleCount = tb.getInt(1, i);
+      } else if (col1Type == Type.LONG_TYPE) {
+        tupleCount = (int) tb.getLong(1, i);
+      } else {
+        throw new DbException("TupleCount must be of type INT or LONG");
+      }
+      Preconditions.checkState(tupleCount >= 0,
+          "Worker cannot have a negative TupleCount: %s", tupleCount);
+
+      int actualTupleCount = tupleCount;
+      if (hasActualTupleCount) {
+        if (col2Type == Type.INT_TYPE) {
+          actualTupleCount = tb.getInt(2, i);
+        } else if (col2Type == Type.LONG_TYPE) {
+          actualTupleCount = (int) tb.getLong(2, i);
+        } else {
+          throw new DbException("ActualTupleCount must be of type INT or LONG");
+        }
+        Preconditions.checkState(tupleCount >= 0,
+            "Worker cannot have a negative ActualTupleCount: %d",
+            actualTupleCount);
+      }
+
+      WorkerInfo wInfo = new WorkerInfo(tupleCount, actualTupleCount);
+      workerInfo.put(workerID, wInfo);
+      totalTupleCount += tupleCount;
+    }
+  }
+
+  /**
+   * Creates a WithoutReplacement distribution across the workers.
+   *
+   * @param workerInfo
+   *          reference to the workerInfo to modify.
+   * @param totalTupleCount
+   *          total # of tuples across all workers.
+   * @param sampleSize
+   *          total # of samples to distribute across the workers.
+   */
+  private void withReplacementDistribution(
+      TreeMap<Integer, WorkerInfo> workerInfo, int totalTupleCount,
+      int sampleSize) {
+    for (int i = 0; i < sampleSize; i++) {
+      int sampleTupleIdx = rand.nextInt(totalTupleCount);
+      // Assign this tuple to the workerID that holds this sampleTupleIdx.
+      int tupleOffset = 0;
+      for (Map.Entry<Integer, WorkerInfo> iWorker : workerInfo.entrySet()) {
+        WorkerInfo wInfo = iWorker.getValue();
+        if (sampleTupleIdx < wInfo.tupleCount + tupleOffset) {
+          wInfo.sampleSize += 1;
+          break;
+        }
+        tupleOffset += wInfo.tupleCount;
+      }
+    }
+  }
+
+  /**
+   * Creates a WithoutReplacement distribution across the workers.
+   *
+   * @param workerInfo
+   *          reference to the workerInfo to modify.
+   * @param totalTupleCount
+   *          total # of tuples across all workers.
+   * @param sampleSize
+   *          total # of samples to distribute across the workers.
+   */
+  private void withoutReplacementDistribution(
+      TreeMap<Integer, WorkerInfo> workerInfo, int totalTupleCount,
+      int sampleSize) {
+    Map<Integer, Integer> logicalTupleCounts = new TreeMap<>();
```

Comment:
If you're depending on ordered iteration of keys, I think this should be declared as `SortedMap`, even if you only use the `Map` interface.